domain/port/service/service.go

```diff
@@ -78,14 +81,40 @@ func (s *Service) GetUnitOpenedPorts(ctx context.Context, unitUUID coreunit.UUID
 	return s.st.GetUnitOpenedPorts(ctx, unitUUID)
 }
 
-// GetMachineOpenedPorts returns the opened ports for all the units on the machine.
-// Opened ports are grouped first by unit name and then by endpoint.
+// GetMachineOpenedPorts returns the opened ports for all endpoints, for all the
+// units on the machine. Opened ports are grouped first by unit name and then by
+// endpoint.
 //
 // TODO: Once we have a core static machine uuid type, use it here.
 func (s *Service) GetMachineOpenedPorts(ctx context.Context, machineUUID string) (map[string]network.GroupedPortRanges, error) {
 	return s.st.GetMachineOpenedPorts(ctx, machineUUID)
 }
 
+// GetMachineOpenedPortsAndSubnets returns the opened port ranges, including the
+// subnet CIDRs they're open to, for each endpoint, for each unit on a machine.
+//
+// NOTE(jack-w-shaw): This method is a stub implementation, that always returns
+// nil for the endpoint SubnetCIDRs
+// TODO: Once endpoint bindings have been implemented into a DQLite domain, implement
+// this method properly to return the subnet CIDRs each endpoint is open to.
+func (s *Service) GetMachineOpenedPortsAndSubnets(ctx context.Context, machineUUID string) (map[string]map[string]port.PortRangesOnSubnet, error) {
```

Comment:
**todo:** I am not really a big fan to have this kind of very complex map as a result.

Please consider to introduce a structured type to return only a level of map, maximum.

It will force us to provide clean name for whatever we return, but would be far more easier to maintain. map of map shouldn't be returned (or even passed as parameter. They shouldn't exists except in deep implementation details.)

Reply:
Yes this is fair. In other places we return a map of `network.GroupedPortRanges` (i.e. map[string]network.GroupedPortRanges`), which under the surface is a map to a map.

In this case, we're doing something very similar, but attaching collections of SubnetCIDRs to our collections of port rnages.

So I will add a new type, equivalent to GroupedPortRanges, and return a map targeting that, so this method has symmetry with others. I think that will be the most readable option